rust/holochain_bootstrap_core/src/types.rs

```diff
@@ -0,0 +1,155 @@
+    }
+}
+
+impl From<&String> for BCoreError {
+    fn from(s: &String) -> Self {
+        s.to_string().into()
+    }
+}
+
+impl From<&str> for BCoreError {
+    fn from(s: &str) -> Self {
+        s.to_string().into()
+    }
+}
+
+impl BCoreError {
+    /// promote a custom error type to a BCoreError::EDecode
+    pub fn decode(e: impl Into<Box<dyn std::error::Error + Send + Sync>>) -> Self {
+        BCoreError::EDecode(e.into().to_string())
+    }
+
+    /// promote a custom error type to a BCoreError::EOther
+    pub fn other(e: impl Into<Box<dyn std::error::Error + Send + Sync>>) -> Self {
+        BCoreError::EOther(e.into().to_string())
+    }
+}
+
+impl std::fmt::Debug for BCoreError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use BCoreError::*;
+        match self {
+            EBadPubKey => f.write_str("EBadPubKey"),
+            EBadSig => f.write_str("EBadSig"),
+            EBadOp { method, op } => {
+                f.write_str("EBadOp: ")?;
+                f.write_str(method)?;
+                f.write_str("/")?;
+                f.write_str(op)
+            }
+            EDecode(err) => {
+                f.write_str("EDecode: ")?;
+                f.write_str(err)
+            }
+            EOther(oth) => {
+                f.write_str("EOther: ")?;
+                f.write_str(oth)
+            }
+        }
+    }
+}
+
+impl std::fmt::Display for BCoreError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Debug::fmt(self, f)
+    }
+}
+
+impl std::error::Error for BCoreError {}
+
+/// build a BCoreError::EOther from `format!()`-style parameters
+#[macro_export]
+macro_rules! bcore_err {
+    ($($arg: tt)*) => {{
+        $crate::types::BCoreError::EOther(::std::format!($($arg)*))
+    }};
+}
+
+/// Bootstrap Core Result Type
+pub type BCoreResult<T> = std::result::Result<T, BCoreError>;
+
+/// Future type for trait declarations
+pub type BCoreFut<'a, T> = std::pin::Pin<Box<dyn Future<Output = T> + 'a>>;
+
+/// Helper fn to generate a BCoreFut type
+pub fn bcore_fut<'a, R, F: Future<Output = R> + 'a>(f: F) -> BCoreFut<'a, R> {
+    Box::pin(f)
+}
+
+/// HTTP Response Object
+#[derive(Debug)]
+pub struct HttpResponse {
+    /// the status code
+    pub status: u16,
+
+    /// the list of headers to send
+    pub headers: Vec<(String, String)>,
+
+    /// the body content
+    pub body: Vec<u8>,
+}
+
+/// Trait representing a KV implementation
+pub trait AsKV: 'static {
+    /// put data into the KV
+    fn put<'a>(&'a self, key: &str, value: &[u8], ttl_secs: f64) -> BCoreFut<'a, BCoreResult<()>>;
+
+    /// get data from the KV
+    fn get<'a>(&'a self, key: &str) -> BCoreFut<'a, BCoreResult<Box<[u8]>>>;
+
+    /// delete a key from the KV
+    fn delete<'a>(&'a self, key: &str) -> BCoreFut<'a, BCoreResult<()>>;
+
+    /// list keys from the KV
+    fn list<'a>(&self, prefix: Option<&str>) -> BCoreFut<'a, BCoreResult<Vec<String>>>;
```

Comment:
do we want `random`? that allows to randomly get agents without doing pagination over the wire, or restricting random results that are biased to the first page only?

Reply:
This is just an FFI to the cloudflare KV api, when we get around to migrating your typescript functions into rust/wasm, I was imagining keeping your exact API, so yes, doing the random thing.